### PR TITLE
Anonymous Sessions

### PR DESCRIPTION
**Pull request by rwaldron**
https://youtu.be/J-bOe87wphQ

### CODE REVIEW

File: client/components/CreateAccount/CreateAnonymousAccount.jsx [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        this.generateUnusedAnonymousUsername();
+    }
+
+    async generateUnusedAnonymousUsername() {
+        const username = anonymousUsername();
+        try {
+            await fetch(`/api/auth/signup/usernames/${username}/exists`);
+        } catch (error) {
+            if (error) {
+                return await this.generateUnusedAnonymousUsername();
```
Comment (ktbee) [inline client/components/CreateAccount/CreateAnonymousAccount.jsx]: Maybe we should confirm that this error is that the username exists? If it's an unrelated server error, I could see us getting caught in a re-try loop.

File: client/components/CreateAccount/CreateAnonymousAccount.jsx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            // after signup and session create
+            location.href = from ? from.pathname : '/';
+        }
+    }
+
+    onClick(event, { name }) {
+        if (name === 'no') {
+            this.generateUnusedAnonymousUsername();
+        }
+        if (name === 'nvm') {
```
Comment (ktbee) [inline client/components/CreateAccount/CreateAnonymousAccount.jsx]: 'lol'

File: client/components/CreateAccount/CreateAnonymousAccount.jsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            location.href = '/';
+        }
+    }
+
+    render() {
+        const { username } = this.state;
+        const { onClick, onSubmit } = this;
+        const header = username
+            ? 'Do you like this temporary user name?'
```
Comment (ktbee) [inline client/components/CreateAccount/CreateAnonymousAccount.jsx]: I guess what makes this account "temporary" is that it doesn't have a password associated with it?

File: client/components/Login/ConfirmableLogoutMenuItem.jsx [hunk 4]
```diff
@@ -27,17 +28,20 @@ class ConfirmableLogoutMenuItem extends Component {
...
+        const { username } = Session.isSessionActive()
+            ? Session.getSession()
+            : {};
 
         return (
             <React.Fragment>
                 <Menu.Item
                     className="confirmablelogoutmenuitem__anchor-style"
-                    content="Log out"
+                    content={`Log out (${username})`}
```
Comment (ktbee) [inline client/components/Login/ConfirmableLogoutMenuItem.jsx]: Really nice touch adding the username 👌

File: client/components/ScenariosList/index.jsx [hunk 5]
```diff
@@ -23,7 +23,7 @@ const ScenarioEntries = ({ scenarios, isLoggedIn }) => {
                         {description}
                     </Card.Description>
                 </Card.Content>
-                {!deleted_at && isLoggedIn && (
+                {!deleted_at && (
```
Comment (ktbee) [inline client/components/ScenariosList/index.jsx]: Good fix thank you 👍

File: server/service/auth/index.js [hunk 6]
```diff
@@ -1,39 +1,46 @@
...
     validateRequestBody,
     validateRequestUsernameAndEmail,
     checkForDuplicate,
     createUser,
     respondWithUser
 ]);
 
-authRouter.post('/login', [
+router.get('/signup/usernames/:username/exists', [
+    checkForDuplicate,
```
Comment (ktbee) [inline server/service/auth/index.js]: Hooray for existing middleware!

File: client/components/CreateAccount/CreateAnonymousAccount.jsx (hunk 2 above)
Comment (rwaldron) [reply client/components/CreateAccount/CreateAnonymousAccount.jsx]: 🤣

File: client/components/Login/ConfirmableLogoutMenuItem.jsx (hunk 4 above)
Comment (rwaldron) [reply client/components/Login/ConfirmableLogoutMenuItem.jsx]: I was getting confused about which user I was 🤣

File: server/service/auth/index.js (hunk 6 above)
Comment (rwaldron) [reply server/service/auth/index.js]: Yeah, big time! This required almost NO changes to the backend.

**Comment by rwaldron** [comment]
@ktbee all review feedback has been addressed!

**Pull request merged**